### ISSUE TITLE
Something wrong in regular expression to get workflow name from host

### ISSUE THREAD

**Issue by tjulym** [faasflow/runtime#1]
The regular expression
`var re = regexp.MustCompile(`(?m)^[^:.]+\s*`)`
in runtime/controller/http/new_request_handler_wrapper.go may lead to wrong workflow name.

For example:
"http" will match the name from "http://127.0.0.1:31112/function/sn-flow" according to the re.

And in practice, "10" wiil match the name when I invoke above workflow, which results in the wrong forwaring 
of async chain.

**Comment by s8sg** [comment]
I agree,The current methodology only works for openfaas function but not other runtime.
I think the runtime who should provide a method to get workflow name. 
```go
runtime.GetWorkflowName()
```

I'm curious how are you using it ?

**Comment by tjulym** [comment]
> 
> 
> I agree,The current methodology only works for openfaas function but not other runtime.
> I think the runtime who should provide a method to get workflow name.
> 
> ```go
> runtime.GetWorkflowName()
> ```
> 
> I'm curious how are you using it ?

As mentioned above, I have a workflow called sn-flow which use async chain provided by faas-flow.
Then I invoke it by sending post request to the default url (http://127.0.0.1:31112/function/sn-flow).

However, the workflow is "10" based on the re, which should be sn-flow (all work well when I set flowname="sn-flow").

I wonder if the target host is like sn-flow.service.xxxx ?

Could u pls tell me how can I get the workspace according to my needs?

**Comment by tjulym** [comment]
Sorry, I have found it's my problem.

I have changed the config direct_functions of OpenFaas gateway, which will invoke functions passing through the provider instead of directly over overlay network by DNS name. Therefore, the host is not what the runtime expect. When the config 
is true, everything is ok.

By the way, it seems faas-flow executes DAG nodes according to the order of adding them to the graph and from the top to the bottom one by one instead of parallel execution. 

For example, if I define node 1-7, and edges (1,2), (2,3), (2,4), (1,5), (1,6), (3,7), (4,7), (5,7), (6,7). 
Then faas-flow will execute nodes 1, 2, 5, 6, 3, 4, 7 one by one (I conclude from logs), rather than execute 2, 5, 6 in parallel.

I wonder whether I misinterpret the design (faas-flow is not parallel execution), or my conclusion is wrong (it is parallel execution)?

**Comment by s8sg** [comment]
It is parallel execution, may I know how many instance of fans-flow is running ?

**Comment by tjulym** [comment]
> 
> 
> It is parallel execution, may I know how many instance of fans-flow is running ?

Only one instance. Do you mean that there are should be multiple instances to implement parallel execution instead of goroutine in a instance?

**Comment by s8sg** [comment]
Actually each node request we are enqueuing each request to a nats queue, in OpenFaaS it gets executed in a separate thread. So technically it should be parallel.

Can you please open a new issue related to the parallel execution

**Comment by dimgiagos44** [comment]
@tjulym Is there any way executing the flow with *directFunctions=false*? This flag is important for load balancing reasons if a function has more than one replicas available for invocations...

**Comment by tjulym** [comment]
> 

Sorry, I don't find the solution.

**Comment by dimgiagos44** [comment]
I tried a hack solution. I created a custom OpenFaas template, exactly like faas-flow but i hard-coded the name of my flow in the go files of the template so as to bypass the source code's bug in name resolution, and all worked fine.

**Comment by TYsonHe** [comment]
> I tried a hack solution. I created a custom OpenFaas template, exactly like faas-flow but i hard-coded the name of my flow in the go files of the template so as to bypass the source code's bug in name resolution, and all worked fine.

can you tell me what template is ? is in your repositories? please,i really need the solution

**Comment by dimgiagos44** [comment]
> > I tried a hack solution. I created a custom OpenFaas template, exactly like faas-flow but i hard-coded the name of my flow in the go files of the template so as to bypass the source code's bug in name resolution, and all worked fine.
> 
> can you tell me what template is ? is in your repositories? please,i really need the solution

You can clone the OpenFaas template and store in a repo like [here](https://github.com/dimgiagos44/faas-flow-versions-templates/tree/master). Edit the source code to match your needs and call this repo as the GoLang module for faas-flow function.. I hope this helps
